File: leetcode-04/solutions/python3/0885.py
class Solution:
    def spiralMatrixIII(self, R: int, C: int, r0: int, c0: int) -> List[List[int]]:
        ans = [[r0, c0]]
        x = 0
        y = 1
        i = 0

        while len(ans) < R * C:
            for j in range(i // 2 + 1):
                r0 += x
                c0 += y
                if 0 <= r0 < R and 0 <= c0 < C:
                    ans.append([r0, c0])
            x, y = y, -x
            i += 1

        return ans

File: leetcode-04/solutions/cpp/0885.cpp
class Solution {
 public:
  vector<vector<int>> spiralMatrixIII(int R, int C, int r0, int c0) {
    vector<vector<int>> ans = {{r0, c0}};
    int x = 0;
    int y = 1;

    for (int i = 0; ans.size() < R * C; i++) {
      for (int j = 0; j < i / 2 + 1; j++) {
        r0 += x;
        c0 += y;
        if (0 <= r0 && r0 < R && 0 <= c0 && c0 < C) ans.push_back({r0, c0});
      }
      swap(x, y);
      y *= -1;
    }

    return ans;
  }
};